smart_open/tests/test_gcs.py

```diff
@@ -0,0 +1,659 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 Radim Rehurek <me@radimrehurek.com>
+#
+# This code is distributed under the terms and conditions
+# from the MIT License (MIT).
+#
+import gzip
+import inspect
+import io
+import logging
+import os
+import time
+import uuid
+import unittest
+try:
+    from unittest import mock
+except ImportError:
+    import mock
+import warnings
+
+import google.cloud
+import google.api_core.exceptions
+import six
+
+import smart_open
+
+BUCKET_NAME = 'test-smartopen-{}'.format(uuid.uuid4().hex)
+BLOB_NAME = 'test-blob'
+WRITE_BLOB_NAME = 'test-write-blob'
+DISABLE_MOCKS = os.environ.get('SO_DISABLE_MOCKS') == "1"
+
+logger = logging.getLogger(__name__)
+
+
+def ignore_resource_warnings():
+    if six.PY2:
+        return
+    warnings.filterwarnings("ignore", category=ResourceWarning, message="unclosed.*<ssl.SSLSocket.*>")  # noqa
+
+
+class FakeBucket(object):
+    def __init__(self, client, name=None):
+        self.client = client  # type: FakeClient
+        self.name = name
+        self.blobs = []
+        self._exists = True
+
+    def blob(self, blob_id):
+        blob = next((x for x in self.blobs if x.name == blob_id), None)
+        if blob is None:
+            blob = FakeBlob(blob_id, self)
+        return blob
+
+    def delete(self):
+        self.client._delete_bucket(self)
+        self._exists = False
+
+    def exists(self):
+        return self._exists
+
+    def get_blob(self, blob_id):
+        blob = next((x for x in self.blobs if x.name == blob_id), None)
+        if blob is None:
+            raise google.cloud.exceptions.NotFound('Blob {} not found'.format(blob_id))
+        return blob
+
+    def list_blobs(self):
+        return self.blobs
+
+    def _delete_blob(self, blob):
+        self.blobs.remove(blob)
+
+
+class FakeBlob(object):
+    RESUMABLE_SESSION_URI_TEMPLATE = (
+        'https://www.googleapis.com/upload/storage/v1/b/'
+        '{bucket}'
+        '/o?uploadType=resumable&upload_id='
+        '{upload_id}'
+    )
+
+    def __init__(self, name, bucket, create=True):
+        self.name = name
+        self._bucket = bucket  # type: FakeBucket
+        self._exists = False
+        self.__contents = io.BytesIO()
+
+        if create:
+            self._create_if_not_exists()
+
+    def create_resumable_upload_session(self):
+        resumeable_upload_url = self.RESUMABLE_SESSION_URI_TEMPLATE.format(
```

Comment:
Please don't use .format, we avoid it in our source code. Use `formatstr % dict(foo='bar')`.